npy_complex256 seems to be undeclared on Windows x64

I'm trying to build h5py on Windows x64 and I got the following error message:

``` 
C:\Users\wlee\AppData\Local\Programs\Common\Microsoft\Visual C++ for Python\9.0\VC\Bin\amd64\cl.exe /c /nologo /Ox /MD /W3 /GS- /DNDEBUG -DH5_USE_16_API -D_HDF5USEDLL_ -DH5_BUILT_AS_DYNAMIC_LIB -Ic:\temp\conda_bld\wlee\bld\h5py_1479831517399\_b_env\Library\include -Ic:\temp\conda_bld\wlee\bld\h5py_1479831517399\work\lzf -Ic:\temp\conda_bld\wlee\bld\h5py_1479831517399\work\windows -Ic:\temp\conda_bld\wlee\bld\h5py_1479831517399\_b_env\lib\site-packages\numpy\core\include -Ic:\temp\conda_bld\wlee\bld\h5py_1479831517399\_b_env\include -Ic:\temp\conda_bld\wlee\bld\h5py_1479831517399\_b_env\PC /Tcc:\temp\conda_bld\wlee\bld\h5py_1479831517399\work\h5py\h5t.c /Fobuild\temp.win-amd64-2.7\Release\temp\conda_bld\wlee\bld\h5py_1479831517399\work\h5py\h5t.obj h5t.c
c:\temp\conda_bld\wlee\bld\h5py_1479831517399\_b_env\lib\site-packages\numpy\core\include\numpy\npy_1_7_deprecated_api.h(12) : Warning Msg: Using deprecated NumPy API, disable it by #defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION c:\temp\conda_bld\wlee\bld\h5py_1479831517399\work\h5py\h5t.c(3324) : warning C4
244: 'function' : conversion from 'hsize_t' to 'int', possible loss of data
c:\temp\conda_bld\wlee\bld\h5py_1479831517399\work\h5py\h5t.c(14433) : error C2065: 'npy_complex256' : undeclared identifier
c:\temp\conda_bld\wlee\bld\h5py_1479831517399\work\h5py\h5t.c(14442) : error C2065: 'npy_complex256' : undeclared identifier
c:\temp\conda_bld\wlee\bld\h5py_1479831517399\work\h5py\h5t.c(14442) : error C2059: syntax error : ')'
c:\temp\conda_bld\wlee\bld\h5py_1479831517399\work\h5py\h5t.c(14451) : error C2065: 'npy_complex256' : undeclared identifier
c:\temp\conda_bld\wlee\bld\h5py_1479831517399\work\h5py\h5t.c(14451) : error C2059: syntax error : ')'
error: command 'C:\\Users\\wlee\\AppData\\Local\\Programs\\Common\\Microsoft\\Visual C++ for Python\\9.0\\VC\\Bin\\amd64\\cl.exe' failed with exit status 2
```

I'm using the latest code from master with numpy 1.11 and python 2.7.  The same build on Linux seems to be work ok.  Any pointer on where I can look at this issue is greatly appreciated.

Thanks.

--- a/h5py/api_compat.h
+++ b/h5py/api_compat.h
@@ -30,13 +30,19 @@ typedef void *PyMPI_MPI_Message;
 
 #define h5py_size_n64 (sizeof(npy_complex64))
 #define h5py_size_n128 (sizeof(npy_complex128))
+
+#ifdef NPY_COMPLEX256
 #define h5py_size_n256 (sizeof(npy_complex256))
+#endif
 
 #define h5py_offset_n64_real (HOFFSET(npy_complex64, real))
 #define h5py_offset_n64_imag (HOFFSET(npy_complex64, imag))
 #define h5py_offset_n128_real (HOFFSET(npy_complex128, real))
 #define h5py_offset_n128_imag (HOFFSET(npy_complex128, imag))
+
+#ifdef NPY_COMPLEX256
 #define h5py_offset_n256_real (HOFFSET(npy_complex256, real))
 #define h5py_offset_n256_imag (HOFFSET(npy_complex256, imag))
+#endif
 
 #endif

--- a/h5py/api_types_ext.pxd
+++ b/h5py/api_types_ext.pxd
@@ -63,13 +63,15 @@ cdef extern from "api_compat.h":
 
   size_t h5py_size_n64
   size_t h5py_size_n128
-  size_t h5py_size_n256
   size_t h5py_offset_n64_real
   size_t h5py_offset_n64_imag
   size_t h5py_offset_n128_real
   size_t h5py_offset_n128_imag
-  size_t h5py_offset_n256_real
-  size_t h5py_offset_n256_imag
+
+  IF COMPLEX256_SUPPORT:
+    size_t h5py_size_n256
+    size_t h5py_offset_n256_real
+    size_t h5py_offset_n256_imag
 
 cdef extern from "lzf_filter.h":
 

--- a/h5py/h5t.pyx
+++ b/h5py/h5t.pyx
@@ -16,6 +16,7 @@
 """
 
 # Pyrex compile-time imports
+include "config.pxi"
 from _objects cimport pdefault
 
 from numpy cimport dtype, ndarray
@@ -1394,10 +1395,13 @@ cdef TypeCompoundID _c_complex(dtype dt):
             tid_sub = H5T_NATIVE_DOUBLE
 
     elif length == 32:
-        size = h5py_size_n256
-        off_r = h5py_offset_n256_real
-        off_i = h5py_offset_n256_imag
-        tid_sub = H5T_NATIVE_LDOUBLE
+        IF COMPLEX256_SUPPORT:
+            size = h5py_size_n256
+            off_r = h5py_offset_n256_real
+            off_i = h5py_offset_n256_imag
+            tid_sub = H5T_NATIVE_LDOUBLE
+        ELSE:
+            raise TypeError("Illegal length %d for complex dtype" % length)
     else:
         raise TypeError("Illegal length %d for complex dtype" % length)
 

--- a/h5py/tests/old/test_dataset.py
+++ b/h5py/tests/old/test_dataset.py
@@ -93,6 +93,7 @@ class TestCreateShape(BaseDataset):
         dset = self.f.create_dataset('foo', (63,), dtype=np.longdouble)
         self.assertEqual(dset.dtype, np.longdouble)
 
+    @ut.skipIf(not hasattr(np, "complex256"), "No support for complex256")
     def test_complex256(self):
         """ Confirm that the default dtype is float """
         dset = self.f.create_dataset('foo', (63,),

--- a/setup_build.py
+++ b/setup_build.py
@@ -145,6 +145,7 @@ class h5py_build_ext(build_ext):
         """ Distutils calls this method to run the command """
 
         from Cython.Build import cythonize
+        import numpy
 
         # Provides all of our build options
         config = self.distribution.get_command_obj('configure')
@@ -176,10 +177,14 @@ DEF MPI4PY_V2 = %(mpi4py_v2)s
 DEF HDF5_VERSION = %(version)s
 DEF SWMR_MIN_HDF5_VERSION = (1,9,178)
 DEF VDS_MIN_HDF5_VERSION = (1,9,233)
+DEF COMPLEX256_SUPPORT = %(complex256_support)s
 """
-                s %= {'mpi': bool(config.mpi),
-                      'mpi4py_v2': bool(v2),
-                      'version': tuple(int(x) for x in config.hdf5_version.split('.'))}
+                s %= {
+                    'mpi': bool(config.mpi),
+                    'mpi4py_v2': bool(v2),
+                    'version': tuple(int(x) for x in config.hdf5_version.split('.')),
+                    'complex256_support': hasattr(numpy, 'complex256')
+                }
                 s = s.encode('utf-8')
                 f.write(s)
 

